### PR TITLE
Add typing for letrec definition

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Solves #278

### CODE REVIEW

**Comment by stefanobaghino-da** [comment]
CI reports a failure but I'm not sure of what the error is. I compared the output of running `make test` on `master` and my branch and they seem identical.

**Comment by stefanobaghino-da** [comment]
@phadej Do you have some suggestion?

**Comment by stefanobaghino-da** [comment]
Hey @phadej, let me know if you can help, otherwise I can try to find out what the issue is. I'd love to contribute to this library.

**Comment by StoneCypher** [comment]
if the problem is that david's detecting out of date deps, probably just bump the dep in the pr

**Comment by stefanobaghino-da** [comment]
I see, I was surprised by the issue because it didn't appear on `master`, so I imagine `master` was fairly stale and this caused the misalignment. I upgraded the dependencies as required by `david`.

File: lib/jsverify.d.ts [hunk 1]
```diff
@@ -114,6 +114,7 @@ declare namespace JSVerify {
   function oneof<A, B, C, D, E>(gs: [Arbitrary<A>, Arbitrary<B>, Arbitrary<C>, Arbitrary<D>, Arbitrary<E>]): Arbitrary<A | B | C | D | E>;
   function oneof<T>(gs: Arbitrary<T>[]): Arbitrary<T>;
   function record<T>(arbs: { [P in keyof T]: Arbitrary<T[P]> }): Arbitrary<T>;
+  function letrec<A>(tie: (f: (key: string) => Arbitrary<A>) => { [key: string]: Arbitrary<A> }): { [key: string]: Arbitrary<A> }
```
Comment (phadej) [inline lib/jsverify.d.ts]: doesn't typescript allow narrowing down `string` to only keys used in the record? I haven't followed recent developments, but recall something like that? (e.g. as in `record` above)?

Comment (stefanobaghino-da) [reply lib/jsverify.d.ts]: Very good remark, thanks. I just pushed a new commit to fix this.

Comment (qlonik) [reply lib/jsverify.d.ts]: You could simplify it even further, if you say that the `K extends string`:

```suggestion
 function letrec<K extends string, V>(tie: (mapping: (key: K) => Arbitrary<V>) => Record<K, Arbitrary<V>>): Record<K, Arbitrary<V>>;
```

**Comment by StoneCypher** [comment]
force push 👎

**Comment by phadej** [comment]
Force pushes are fine by me, it's not master or release branch

**Comment by StoneCypher** [comment]
oh, my misread.  apologies

**Comment by stefanobaghino-da** [comment]
If there's anything more I can do for this PR, I'd be happy to do it. :slightly_smiling_face:

**Comment by phadej** [comment]
@stefanobaghino-da thanks

**Pull request merged**